src/smartpeak/source/ui/SequenceGroupsEditorWidget.cpp

```diff
@@ -0,0 +1,122 @@
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL ANY OF THE AUTHORS OR THE CONTRIBUTING
+// INSTITUTIONS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Douglas McCloskey, Bertrand Boudaud $
+// $Authors: Bertrand Boudaud $
+// --------------------------------------------------------------------------
+
+#include <SmartPeak/ui/SequenceGroupsEditorWidget.h>
+#include <misc/cpp/imgui_stdlib.h>
+
+namespace SmartPeak
+{
+
+  void SequenceGroupsEditorWidget::open(const std::set<std::string>& choices, 
+                                        const std::string& current_choice, 
+                                        std::function<void(const std::string&)> ok_callback)
+  {
+    action_choice_ = EActionChoice_MoveSegment;
+    ok_callback_ = ok_callback;
+    groups_ = choices;
+    new_group_.clear();
+    current_choice_ = current_choice;
+    ImGui::OpenPopup(title_.c_str());
+  }
+
+  void SequenceGroupsEditorWidget::draw()
+  {
+    float popup_width = 400;
+
+    if (!ImGui::BeginPopupModal(title_.c_str(), NULL, ImGuiWindowFlags_AlwaysAutoResize)) {
+      return;
+    }
+
+    // one documented way to set popup width
+    const auto cursor_pos = ImGui::GetCursorPosX();
+    ImGui::SetCursorPosX(popup_width);
+    ImGui::SetCursorPosX(cursor_pos);
+
+    ImGui::RadioButton(move_option_message_.c_str(), &action_choice_, EActionChoice_MoveSegment);
+    ImGui::SameLine();
+    ImGui::RadioButton(create_option_message_.c_str(), &action_choice_, EActionChoice_CreateSegment);
+
+    if (action_choice_ == EActionChoice_CreateSegment)
+    {
+      ImGui::InputText(move_action_message_.c_str(), &new_group_);
+    }
+
+    if ((action_choice_ == EActionChoice_MoveSegment))
+    {
+      if (!groups_.empty() && ImGui::BeginCombo(create_action_message_.c_str(), current_choice_.c_str()))
+      {
+        for (const auto valid_string : groups_)
+        {
+          if (ImGui::Selectable(valid_string.c_str()))
+          {
+            current_choice_ = valid_string;
+          }
+        }
+        ImGui::EndCombo();
+      }
+    }
+
+    static const std::vector<char> forbiden_chars = { ';',',' };
+    bool new_group_name_is_valid = !std::any_of(forbiden_chars.begin(), forbiden_chars.end(),
```

Comment:
Nice check.

Reply:
by the way, do we allow one of these editable fields to be empty (I guess at least the replicate_group_name will be)?

Reply:
They can be empty strings.  That could serve as the "default" grouping.